src/main/java/uk/gov/hmcts/darts/courthouse/service/CourthouseServiceImpl.java

```diff
@@ -121,6 +135,150 @@ public CourthouseEntity addCourtHouse(Courthouse courthouse) {
         return courthouseRepository.saveAndFlush(mappedEntity);
     }
 
+    private final SecurityGroupIdMapper securityGroupIdMapper;
+
+    @Override
+    public ExtendedCourthousePost mapFromEntitiesToExtendedCourthousePost(CourthousePost courthousePost) {
+        final SecurityRoleEntity securityRoleTranscriber = getSecurityRoleByRoleName("TRANSCRIBER");
+        final SecurityRoleEntity securityRoleRequester = getSecurityRoleByRoleName("REQUESTER");
+        final SecurityRoleEntity securityRoleApprover = getSecurityRoleByRoleName("APPROVER");
+
+        //check duplicate names
+        checkCourthouseNameIsUnique(courthousePost);
+        checkCourthouseDisplayNameIsUnique(courthousePost);
+
+        List<Integer> securityGroupIds = (courthousePost.getSecurityGroupIds() == null) ? new ArrayList<>() : courthousePost.getSecurityGroupIds();
+        courthousePost.securityGroupIds(securityGroupIds);
+
+        //for each security_group get role_ids
+        List<Integer> securityRoleIds = new ArrayList<>();
+        for (Integer secGrpIds : securityGroupIds) {
+            Optional<SecurityGroupEntity> securityGroupEntity = securityGroupRepository.findById(secGrpIds);
+            securityGroupEntity.ifPresent(groupEntity -> securityRoleIds.add(groupEntity.getSecurityRoleEntity().getId()));
+        }
+
+        //ensure if security roles exist, TRANSCRIBER role also exists
+        if (securityRoleTranscriber != null && (!securityRoleIds.isEmpty() && !securityRoleIds.contains(securityRoleTranscriber.getId()))) {
+                throw new DartsApiException(AuthorisationError.BAD_REQUEST_ONLY_TRANSCIBER_ROLES_MAY_BE_ASSIGNED);
+
+        }
+
+        CourthouseEntity mappedEntity = getMappedEntityWithRegion(courthousePost);
+
+        Set<SecurityGroupEntity> securityGroupEntities = getSecurityGroupEntities(courthousePost, mappedEntity,
+                                                                                  securityRoleIds, securityRoleRequester,
+                                                                                  securityRoleApprover);
+        mappedEntity.setSecurityGroups(securityGroupEntities);
+
+        courthouseRepository.saveAndFlush(mappedEntity);
+
+        return getExtendedCourthousePost(courthousePost, mappedEntity);
+    }
+
+    private ExtendedCourthousePost getExtendedCourthousePost(CourthousePost courthousePost, CourthouseEntity mappedEntity) {
```

Comment:
Readability would be improved by renaming this method to better reflect what it's doing, e.g. `mapToResponse`.

I think it's dangerous to accept `CourthousePost courthousePost` into this method, as it could make the response misleading. The response should be based purely on what now exists in the DB, rather than what was provided in the request.

Reply:
method name updated